xbmc/pvr/PVRNumericChannelInputHandler.cpp

```diff
@@ -0,0 +1,159 @@
+/*
+ *      Copyright (C) 2012-2017 Team Kodi
+ *      http://kodi.tv
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "input/Key.h"
+#include "messaging/ApplicationMessenger.h"
+
+#include "pvr/PVRManager.h"
+#include "pvr/channels/PVRChannelGroupsContainer.h"
+
+#include "PVRNumericChannelInputHandler.h"
+
+using namespace KODI::MESSAGING;
+
+namespace PVR
+{
+
+CPVRNumericChannelInputHandler::CPVRNumericChannelInputHandler(int iDigits, int iSwitchDelay)
+: m_iDigits(iDigits),
+  m_iSwitchDelay(iSwitchDelay),
+  m_timer(this)
+{
+}
+
+void CPVRNumericChannelInputHandler::OnTimeout()
+{
+  int iChannelNumber;
+  {
+    CSingleLock lock(m_critSection);
+    iChannelNumber = GetChannel();
+    ResetChannel();
+  }
+
+  SwitchToChannel(iChannelNumber);
+}
+
+void CPVRNumericChannelInputHandler::AppendChannelDigit(int iDigit)
+{
+  if (iDigit < 0 || iDigit > 9)
+    return;
+
+  CSingleLock lock(m_critSection);
+
+  if (iDigit == 0 && m_digits.size() == 0)
+  {
+    // special case. if first input is zero, switch to previously played channel.
+    lock.Leave();
+    SwitchToPreviousChannel();
+  }
+  else
+  {
+    if (m_digits.size() == m_iDigits)
+      m_digits.pop_front();
+
+    m_digits.emplace_back(iDigit);
+
+    // recalc channel string
+    m_strChannel.erase();
+    if (m_digits.size() != m_iDigits || GetChannel() > 0)
+    {
+      for (int digit : m_digits)
+        m_strChannel.append(std::to_string(digit));
```

Comment:
ugh. will jenkins let find out and find an alternative if it turns ot that you are right.